tracing-attributes/src/lib.rs

```diff
@@ -579,3 +668,88 @@ fn instrument_err(args: &[NestedMeta]) -> bool {
         _ => false,
     })
 }
+
+// Get the name of the inner function we need to hook, if the function was generated by async-trait.
+// When we are given a function generated by async-trait, that function in only a "temporary"
+// function that returns a pinned future, and it is that pinned function that needs to be
+// instrumened, otherwise we will only collect information on the moment the future was "built",
+// and not it trues span of execution.
+// So we inspect the block of the function to find if we can find the pattern
+// `async fn foo<...>(...) {...}; Box::pin(foo<...>(...))` and return the name `foo` if that
+// is the case. Our caller will then be able to use that information to instrument the proper
+// function.
+// (this follows the approach suggested in
+// https://github.com/dtolnay/async-trait/issues/45#issuecomment-571245673)
+fn get_async_trait_name(block: &Block, block_is_async: bool) -> Option<String> {
+    // are we in an async context ? If yes, this isn't a async_trait-like pattern
+    if block_is_async {
+        return None;
+    }
+
+    // list of async functions declared inside the block
+    let mut inside_funs = Vec::new();
+    // last expression declared in the block (it determines the return value of the block, so
+    // that if we are working on a function whose `trait` or `impl` declaration is annotated by
+    // async_trait, this is quite likely the point where the future is pinned)
+    let mut last_expr = None;
+
+    // obtain the list of direct internal functions and the last expression of the block
+    for stmt in &block.stmts {
+        if let Stmt::Item(Item::Fn(fun)) = &stmt {
+            // is the function declared as async ? If so, this is a good candidate, let's keep it
+            // in hand
+            if fun.sig.asyncness.is_some() {
+                inside_funs.push(fun.sig.ident.to_string());
+            }
+        } else if let Stmt::Expr(e) = &stmt {
+            last_expr = Some(e);
+        }
+    }
+
+    // let's play with (too much) pattern matching
+    // is the last expression a function call ?
+    if let Some(Expr::Call(ExprCall {
+        func: outside_func,
+        args: outside_args,
+        ..
+    })) = last_expr
+    {
+        if let Expr::Path(path) = outside_func.as_ref() {
+            // is it a call to `Box::pin()` ?
+            let mut merged_path = String::new();
+            for i in 0..path.path.segments.len() {
+                merged_path.push_str(&path.path.segments[i].ident.to_string());
```

Comment:
What we're doing here is writing the ident's `fmt::Display` output to the end of `path_str`. Using `to_string()` creates a new `String`, which allocates memory for a new string that we then immediately throw away. What if we used `write!` instead, like this:

```suggestion
                use std::fmt::Write;
                write!(&mut merged_path, "{}", &path.path.segments[i].ident.to_string())
                    .expect("writing to a String never fails");
```